nilearn/glm/tests/test_bids.py

```diff
@@ -0,0 +1,202 @@
+
+
+def test_first_level_from_bids_validation_input():
+    with InTemporaryDirectory():
+        bids_path = create_fake_bids_dataset(n_sub=3,
+                                             n_ses=2,
+                                             tasks=['localizer', 'main'],
+                                             n_runs=[1, 3])
+        # test arguments are provided correctly
+        with pytest.raises(TypeError):
+            first_level_from_bids(2,
+                                  task_label='main',
+                                  space_label='MNI')
+        with pytest.raises(ValueError):
+            first_level_from_bids('lolo',
+                                  task_label='main',
+                                  space_label='MNI')
+        with pytest.raises(TypeError):
+            first_level_from_bids(bids_path,
+                                  task_label=2,
+                                  space_label='MNI')
+        with pytest.raises(TypeError):
+            first_level_from_bids(bids_path,
+                                  task_label='main',
+                                  space_label='MNI',
+                                  model_init=[])
+
+        with pytest.raises(TypeError,
+                           match="space_label must be a string"):
+            first_level_from_bids(bids_path,
+                                  task_label='main',
+                                  space_label=42)
+
+        with pytest.raises(TypeError,
+                           match="img_filters must be a list"):
+            first_level_from_bids(bids_path,
+                                  task_label='main',
+                                  img_filters="foo")
+        with pytest.raises(TypeError,
+                           match="filters in img"):
+            first_level_from_bids(bids_path,
+                                  task_label='main',
+                                  img_filters=[(1, 2)])
+        with pytest.raises(ValueError,
+                           match="field foo is not a possible filter."):
+            first_level_from_bids(bids_path,
+                                  task_label='main',
+                                  img_filters=[("foo", "bar")])
+
+
+def test_first_level_from_bids_with_missing_files():
+    with InTemporaryDirectory():
+        bids_path = create_fake_bids_dataset(n_sub=3,
+                                             n_ses=2,
+                                             tasks=['localizer', 'main'],
+                                             n_runs=[1, 3])
+        # test repeated run tag error when run tag is in filenames
+        # can arise when desc or space is present and not specified
+        #
+        # desc not specified
+        with pytest.raises(ValueError):
+            first_level_from_bids(bids_path,
+                                  task_label='main',
+                                  space_label='T1w')
```

Comment:
yup I have already started breaking the tests into actual independent tests because I was getting unhelpful tests failures.

will finish breaking those up

good idea for the fixture

Reply:
did not use a fixture but ending up refactoring, splitting the test or parametrazing as much as I could.